docs/examples/using-terraform-plugin.md

```diff
@@ -0,0 +1,265 @@
+- id: tfrun
+  image: vorteil/terraform:v1
+  files:
+  - key: terraform-examples
+    scope: instance
+    type: tar.gz
+description: Clones a repository and uses terraform to deploy an instance
+states:
+  # continued in next code block
+```
+
+**NOTE: The files attribute is empty and will be filled in via the git state.**
+
+## Git
+The following state fetches the repository and clones it into an instance variable to be used in the terraform container.
+
+```yaml
+  - id: cloning
+    type: action
+    action:
+      function: git
+      input: 
+        cmds: ["clone https://github.com/vorteil/terraform-examples.git $out/instance/terraform-examples"]
+    transition: deploy_{CLOUD}
+    log: .
+```
+
+## Terraform
+The terraform files are provided from the git repository. We just need to add in the secret variables to spawn the machines.
+
+The execution-folder directs the Terraform container to tell it where to execute from as all the terraform files are stored in different directories on this repository.
+
+**NOTE: Terraform container has its own http backend located at 'http://localhost:8001/{state-name}'. If provided args-on-init and state-name we will write the tfstate to a workflow variable.**
+
+### Azure
+The only secrets required to run with Azure is a subscription_id, client_id, client_secret and a tenant_id.
+
+
+```yaml
+  - id: deploy_azure
+    type: action
+    log: .
+    action:
+        secrets: ["CLIENT_ID", "TENANT_ID", "SUBSCRIPTION_ID", "CLIENT_SECRET"]
+        function: tfrun
+        input: 
+            action: "{{.action}}"
+            args-on-init: ["-backend-config=address=http://localhost:8001/terraform-azure-instance"]
+            execution-folder: terraform-examples/azure
+            variables:
+                state-name: terraform-azure-instance
+                subscription_id: "{{.secrets.SUBSCRIPTION_ID}}"
+                client_id: "{{.secrets.CLIENT_ID}}"
+                client_secret: "{{.secrets.CLIENT_SECRET}}"
+                tenant_id: "{{.secrets.TENANT_ID}}"
+    transform: |
+      {
+        "ip": .ip
+      }
+    transition: send_message
+```
+
+### Google Cloud Platform
+The only secrets required to run with Google is a service_account_key and a project_id.
+
+```yaml
+  - id: deploy_gcp
+    type: action
+    log: .
+    action:
+      secrets: ["SERVICE_ACCOUNT_KEY", "PROJECT_ID"]
+      function: tfrun
+      input:
+        execution-folder: terraform-examples/google
+        action: "{{.action}}"
+        args-on-init: ["-backend-config=address=http://localhost:8001/terraform-gcp-instance"]
+        variables:
+          state-name: terraform-gcp-instance
+          project_id: "{{.secrets.PROJECT_ID}}"
+          service_account_key: "{{.secrets.SERVICE_ACCOUNT_KEY}}"
+    transform: |
+      {
+          "ip": .ip
+      }
+    transition: send_message
+```
+
+### Amazon Web Services
+The only secrets required to run with Amazon is an access_key and a secret_key.
```

Comment:
The only secrets required to run this workflow with Amazon Web Services are:
- access_key
- secret_key


(Isnt it technically access_key, and access_key_secret?)